File: api/src/main/java/com/gkh/syntheticmonitor/model/Report.java
package com.gkh.syntheticmonitor.model;

import com.fasterxml.jackson.annotation.JsonGetter;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Access(AccessType.FIELD)
public class Report {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	private String name;
	
	private transient int numberOfTests;
	private transient boolean passed;
	private transient long sumResponseTime;
	private transient long minResponseTime;
	private transient long maxResponseTime;
	private transient long averageResponseTime;
	private transient int numberOfActions;
	private transient long sumMaxTimeThreshold;
	private transient boolean allResponseTimeUnderMax;
	private transient boolean allStatusCodeMatching;

	@CreationTimestamp
	private Timestamp timestamp;

	@OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER, orphanRemoval = true)
	@JoinColumn(name = "parent_id")
	@Fetch(FetchMode.JOIN)
	private List<ReportDetail> reportDetails = new ArrayList<>();

	@Column
	@Access(AccessType.PROPERTY)
	public int getNumberOfTests() {
		return this.reportDetails.size();
	}

	@Column
	@Access(AccessType.PROPERTY)
	public boolean isPassed() {
		return this.reportDetails
				.stream()
				.filter(each->each.isStatusSuccess())
				.collect(Collectors.toList()).size() == this.getNumberOfTests();
	}

	@Column
	@Access(AccessType.PROPERTY)
	public long getMinResponseTime() {
		return this.reportDetails
				.stream()
				.mapToLong(ReportDetail::getResponseTime)
				.summaryStatistics().getMin();
	}

	@Column
	@Access(AccessType.PROPERTY)
	public long getMaxResponseTime() {
		return this.reportDetails
				.stream()
				.mapToLong(ReportDetail::getResponseTime)
				.summaryStatistics().getMax();
	}

	@Column
	@Access(AccessType.PROPERTY)
	public long getAverageResponseTime() {
		return (long) this.reportDetails
				.stream()
				.mapToLong(ReportDetail::getResponseTime)
				.summaryStatistics().getAverage();
	}

	@Column
	@Access(AccessType.PROPERTY)
	public long getSumResponseTime() {
		return this.reportDetails
				.stream()
				.mapToLong(ReportDetail::getResponseTime)
				.summaryStatistics().getSum();
	}

	@JsonGetter
	public String getType() {
		if (this.reportDetails.size() == 1) {
			return this.reportDetails.get(0).getType();
		} else {
			return "Transactions";
		}
	}

	@Column
	@Access(AccessType.PROPERTY)
	public int getNumberOfActions() {
		return this.reportDetails.size();
	}



	@Column
	@Access(AccessType.PROPERTY)
	public long getSumMaxTimeThreshold() {
		return this.reportDetails.stream().mapToLong(o->o.getMaximumResponseThreshold()).summaryStatistics().getSum();
	}


	@Column
	@Access(AccessType.PROPERTY)
	public boolean isAllResponseTimeUnderMax() {
		return this.reportDetails.stream().allMatch(p->p.isResponseTimeUnderMax());
	}

	@Column
	@Access(AccessType.PROPERTY)
	public boolean isAllStatusCodeMatching() {
		return this.reportDetails.stream().allMatch(p->p.isStatusCodeMatching());
	}




}

File: api/src/main/kotlin/com/gkh/syntheticmonitor/model/Report.kt
package com.gkh.syntheticmonitor.model

import com.fasterxml.jackson.annotation.JsonGetter
import lombok.AllArgsConstructor
import lombok.Data
import lombok.NoArgsConstructor
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import java.sql.Timestamp
import java.util.stream.Collectors
import javax.persistence.*

@Entity
@Access(AccessType.FIELD)
class Report(
        @Id @GeneratedValue(strategy = GenerationType.AUTO) var id: Long? = null,
        var name: String? = null
) {

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var numberOfTests = 0
        get() = reportDetails.size

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var isPassed = false
        get() = reportDetails
                .stream()
                .filter { each: ReportDetail -> each.isStatusSuccess }
                .collect(Collectors.toList()).size == numberOfTests

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var sumResponseTime: Long = 0
        get() = reportDetails
                .stream()
                .mapToLong { obj: ReportDetail -> obj.responseTime }
                .summaryStatistics().sum

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var minResponseTime: Long = 0
        get() = reportDetails
                .stream()
                .mapToLong { obj: ReportDetail -> obj.responseTime }
                .summaryStatistics().min

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var maxResponseTime: Long = 0
        get() = reportDetails
                .stream()
                .mapToLong { obj: ReportDetail -> obj.responseTime }
                .summaryStatistics().max
    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var averageResponseTime: Long = 0
        get() = reportDetails
                .stream()
                .mapToLong { obj: ReportDetail -> obj.responseTime }
                .summaryStatistics().average.toLong()

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var numberOfActions = 0
        get() = reportDetails.size

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var sumMaxTimeThreshold: Long = 0
        get() = reportDetails.stream().mapToLong { o: ReportDetail -> o.maximumResponseThreshold }.summaryStatistics().sum

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var isAllResponseTimeUnderMax = false
        get() = reportDetails.stream().allMatch { p: ReportDetail -> p.isResponseTimeUnderMax }

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var isAllStatusCodeMatching = false
        get() = reportDetails.stream().allMatch { p: ReportDetail -> p.isStatusCodeMatching }

    @CreationTimestamp
    var timestamp: Timestamp? = null

    @OneToMany(cascade = [CascadeType.ALL],
            fetch = FetchType.EAGER,
            orphanRemoval = true)
    @JoinColumn(name = "parent_id")
    @Fetch(FetchMode.JOIN)
    val reportDetails: MutableList<ReportDetail> = arrayListOf()

    @get:JsonGetter
    val type: String?
        get() = if (reportDetails.size == 1) {
            reportDetails[0].type
        }
        else {
            "Transactions"
        }

}